# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [1]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [3]:
[i for i in range(1,51)]


[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50]

### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [17]:
[i for i in range(2,201) if i%2==0]

[2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200]

### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [21]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [25]:
[j for i in a for j in i]

[0.84062117,
 0.48006452,
 0.7876326,
 0.77109654,
 0.44409793,
 0.09014516,
 0.81835917,
 0.87645456,
 0.7066597,
 0.09610873,
 0.41247947,
 0.57433389,
 0.29960807,
 0.42315023,
 0.34452557,
 0.4751035,
 0.17003563,
 0.46843998,
 0.92796258,
 0.69814654,
 0.41290051,
 0.19561071,
 0.16284783,
 0.97016248,
 0.71725408,
 0.87702738,
 0.31244595,
 0.76615487,
 0.20754036,
 0.57871812,
 0.07214068,
 0.40356048,
 0.12149553,
 0.53222417,
 0.9976855,
 0.12536346,
 0.80930099,
 0.50962849,
 0.94555126,
 0.33364763]

### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [66]:
[j for i in a for j in i if j>=0.5]

[0.84062117,
 0.7876326,
 0.77109654,
 0.81835917,
 0.87645456,
 0.7066597,
 0.57433389,
 0.92796258,
 0.69814654,
 0.97016248,
 0.71725408,
 0.87702738,
 0.76615487,
 0.57871812,
 0.53222417,
 0.9976855,
 0.80930099,
 0.50962849,
 0.94555126]

### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [68]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [75]:
list_shadow = []
for i in b: #2nd we follow our logic
    for j in i:#3rd
        for k in j:#4th
            list_shadow.append(k) #1st we get what we look for
list_shadow
#[k for k in j for j in i for i in b]
[k for i in b for j in i for k in j]

[0.55867166,
 0.06210792,
 0.08147297,
 0.82579068,
 0.91512478,
 0.06833034,
 0.05440634,
 0.65857693,
 0.30296619,
 0.06769833,
 0.96031863,
 0.51293743,
 0.09143215,
 0.71893382,
 0.45850679,
 0.58256464,
 0.59005654,
 0.56266457,
 0.71600294,
 0.87392666,
 0.11434044,
 0.8694668,
 0.65669313,
 0.10708681,
 0.07529684,
 0.46470767,
 0.47984544,
 0.65368638,
 0.14901286,
 0.23760688]

### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [79]:
#[j for i in a for j in i]
#[j for i in a for j in i if j>=0.5]

[j[2] for i in b for j in i if j[2]<=0.5] #l'index de l'objet recherché est 2. 
                                        #On se soucis uniquement de l'élément dont on tire l'index


[0.08147297,
 0.06833034,
 0.30296619,
 0.45850679,
 0.11434044,
 0.10708681,
 0.47984544,
 0.23760688]

### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

In [83]:
file_list = [f for f in os.listdir('/home/utilisateur/Documents/data-labs/module-1/lab-list-comprehensions/data') if f.endswith('.csv')]
file_list

['sample_file_4.csv',
 'sample_file_8.csv',
 'sample_file_9.csv',
 'sample_file_7.csv',
 'sample_file_1.csv',
 'sample_file_3.csv',
 'sample_file_0.csv',
 'sample_file_2.csv',
 'sample_file_6.csv',
 'sample_file_5.csv']

### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

In [89]:
data_sets = [pd.read_csv(os.path.join('/home/utilisateur/Documents/data-labs/module-1/lab-list-comprehensions/data', f)) for f in file_list]
data = pd.concat(data_sets, axis=0).head(10)
data

0         1         2         3         4         5         6  \
0  0.901381  0.790944  0.596603  0.354721  0.357002  0.321325  0.738398   
1  0.763823  0.537281  0.708804  0.975269  0.053777  0.957740  0.389137   
2  0.571864  0.386243  0.567476  0.509283  0.552594  0.173807  0.229092   
3  0.956168  0.422274  0.121713  0.685208  0.713370  0.416245  0.337151   
4  0.121826  0.237479  0.553956  0.173660  0.354423  0.532080  0.271361   
0  0.132453  0.106635  0.673568  0.507253  0.162925  0.130042  0.980388   
1  0.384938  0.767604  0.850528  0.605998  0.826172  0.667488  0.142045   
2  0.318691  0.675720  0.807189  0.752518  0.531342  0.703225  0.598293   
3  0.538593  0.291241  0.407187  0.613260  0.851424  0.962086  0.590512   
4  0.713271  0.156210  0.392800  0.702348  0.009400  0.030304  0.960066   

          7         8         9        10        11        12        13  \
0  0.777885  0.775623  0.717817  0.788198  0.962214  0.394465  0.878922   
1  0.689829  0.510885  0.737445  0.097678  0.242906  0.073603  0.497591   
2  0.257945  0.313437  0.418311  0.750007  0.604759  0.234886  0.832337   
3  0.570693  0.435938  0.360098  0.037437  0.167545  0.847880  0.773456   
4  0.766006  0.607020  0.197302  0.526352  0.676571  0.473574  0.953228   
0  0.623167  0.074066  0.111557  0.864664  0.093637  0.446974  0.022525   
1  0.086904  0.912348  0.892183  0.426389  0.776552  0.429496  0.602056   
2  0.921508  0.563333  0.184300  0.667625  0.270017  0.573440  0.163669   
3  0.980700  0.159022  0.684265  0.898633  0.470113  0.756593  0.282408   
4  0.688937  0.260252  0.634608  0.273775  0.264145  0.682941  0.908499   

         14        15        16        17        18        19  
0  0.084206  0.680941  0.992897  0.267455  0.212403  0.572150  
1  0.798467  0.789078  0.947559  0.631497  0.057953  0.060306  
2  0.421263  0.455785  0.893841  0.726412  0.458360  0.048740  
3  0.490628  0.544232  0.954668  0.567280  0.571539  0.560742  
4  0.413186  0.957242  0.642558  0.351496  0.547297  0.177401  
0  0.713369  0.707874  0.563506  0.982256  0.562498  0.968681  
1  0.283769  0.166793  0.316878  0.688337  0.169525  0.966198  
2  0.032852  0.689315  0.973684  0.719610  0.163670  0.724486  
3  0.551369  0.732153  0.494100  0.792803  0.794426  0.837695  
4  0.450950  0.200761  0.968324  0.221165  0.891382  0.398914

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.